### PR TITLE
Make OPENSSL_sk_push return only 0 or 1

### PR DESCRIPTION
**Pull request by t8m**
Most of the callers do not actually check for
the special -1 return condition because they do not pass NULL to it. It is also extremely improbable that any code depends on this -1 return value in this condition so it can be safely changed to 0 return.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by levitte** [comment]
**OTC: is this API change ok?**

**Review by slontis — COMMENTED** [review]
There are == and > 0 tests is a few places.. (So the == ones are broken currently).
It does seem at least inconsistent that the insert returns 0 for the same check..
What about OPENSSL_sk_num() ?

**Comment by t8m** [comment]
> What about OPENSSL_sk_num() ?

I wouldn't change that as it makes sense to distinguish an empty stack from NULL by the return value. IMO that would certainly be an API break as somebody could depend on that -1 return.

**Comment by t8m** [comment]
**OTC: This PR is acceptable for master branch, if it has a CHANGES.md entry.**

File: doc/man3/DEFINE_STACK_OF.pod [hunk 1]
```diff
@@ -297,6 +296,9 @@ and B<sk_I<TYPE>_find_all>() calls are read-only and do not sort the
 stack.  To avoid any performance implications this change introduces,
 B<sk_I<TYPE>_sort>() should be called before these find operations.
 
+Before OpenSSL 3.3.0 B<sk_I<TYPE>_push>() returned -1 if I<sk> was NULL. It
+was changed to return 0 in this condition as for other errors.
```
Comment (kaduk) [inline doc/man3/DEFINE_STACK_OF.pod:300]: I think it would be okay (but probably not required) to say why the change was made (consistency with insert and render non-buggy existing checks for equality).

**Review by kaduk — APPROVED** [review]
The change does what it says it does, and since I have to defer to the OTC on whether to put it in 3.3 vs 4.0, approving.
I do not have a great sense for what kind of risk assessment was done to compare the risk of breaking existing code relying on a distinct check vs the risk of existing buggy code doing an equality check in a situation where NULL might be passed in, though.

**Comment by DDvO** [comment]
> There are == and > 0 tests is a few places.. (So the == ones are broken currently).

IIUC, the motivation of this PR is to adapt the return values to actual partly wrong use.
I'm a little surprised that no caller code had to be changed.
This seems to imply that for each caller, the result check 
either was already flexible enough to cover also the new semantics,
or there was an equality check that was in fact wrong and is now indirectly fixed by adapting the called function.
So shouldn't this PR be considered (also) a bug fix?

**Comment by t8m** [comment]
> > There are == and > 0 tests is a few places.. (So the == ones are broken currently).
> 
> IIUC, the motivation of this PR is to adapt the return values to actual partly wrong use. I'm a little surprised that no caller code had to be changed. This seems to imply that for each caller, the result check either was already flexible enough to cover also the new semantics, or there was a more strict (equality) check that was in fact wrong and is now indirectly fixed by adapting the called function. So shouldn't this PR be considered (also) a bug fix?

The callers which did not check for >0 return were never calling this with NULL stack. This is not a bug fix.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for the reviews.
